Restructure documentation for request package - fixes #178

jenshalm opened the pull request:
This is an almost complete rewrite of the section about request processing as outlined in this ticket: https://github.com/finagle/finch/issues/178.

This is probably difficult to review based on the diff, you can alternatively read the rendered version.

vkostyukov commented on docs.md:
```diff
@@ -98,249 +110,443 @@ val s = RequiredParam("foo")(req)
+    
+val user: RequestReader[User] = (
+  RequiredParam("name") ~
+  RequiredParam("age").as[Int].shouldNot("be less than 18"){ _ < 18 } ~
```
We can use predefined rule `beLessThan(18)` here. Do you mind also rewriting it in DSL-style:  

```scala
RequiredParam("age").as[Int] shouldNot beLessThan(18) ~
```

jenshalm replied:
I'l switch to the predefined rule. But the DSL-style does not compile without adding more parentheses, so I think this syntax is clearer.

vkostyukov commented on docs.md:
```diff
@@ -98,249 +110,443 @@ val s = RequiredParam("foo")(req)
+val address: RequestReader[Address] = (
+  RequiredParam("street") ~
+  RequiredParam("city") ~
+  RequiredParam("postCode").shouldNot("be longer than 5 characters"){ _ > 5 }
```
We can use predefined rule here: `shouldNot beLongerThan(5)`.

vkostyukov commented:
This is amazing! Thanks a lot @jenshalm for this hard work! I will merge this once you're happy with the state of this PR.

jenshalm commented:
I switched to use built-in rules in the two example you commented. I also added a small section mentioning these built-in rules. The only downside is that our examples for custom rules now still mostly show logic you can do with built-in rules. I did not have a quick idea how to rewrite the samples with something not supported out of the box without making it very contrived. So in my view this is good to be merged now and we can rethink those samples later.

vkostyukov commented:
Super cool! I'm merging this.
